compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/OptimizerUtils.java

```diff
@@ -36,28 +46,72 @@ private OptimizerUtils() {
+   * @param dag the IR DAG to observe.
+   * @return a pair of vertex/edge list and the execution property key index.
    */
-  public static Pair<String, Integer> stringToIdAndEPKeyIndex(final String string) {
-    // Formatted into 9 digits: 0:vertex/edge 1-5:ID 5-9:EP Index.
+  public static Pair<List<Object>, Integer> stringToObjAndEPKeyIndex(final String string, final IRDAG dag) {
+    // Formatted into 9 digits: 0:vertex/edge 1-4:ID 5-8:EP Index.
     if (string.length() != 9) {
       throw new InvalidParameterException("The metric data should follow the format of "
         + "[0]: index indicating vertex/edge, [1-4]: id of the component, and [5-8]: EP Key index. Current: " + string);
     }
     final Integer idx = Integer.parseInt(string.substring(0, 1));
     final Integer numericId = Integer.parseInt(string.substring(1, 5));
-    final String id;
+    final List<Object> objectList = new ArrayList<>();
     if (idx == 1) {
-      id = Util.restoreVertexId(numericId);
+      final String id = Util.restoreVertexId(numericId);
+      objectList.add(dag.getVertexById(id));
     } else if (idx == 2) {
-      id = Util.restoreEdgeId(numericId);
+      final String id = Util.restoreEdgeId(numericId);
+      objectList.add(dag.getEdgeById(id));
     } else {
       throw new UnsupportedMethodException("The index " + idx + " cannot be categorized into a vertex or an edge");
     }
-    return Pair.of(id, Integer.parseInt(string.substring(5, 9)));
+    return Pair.of(objectList, Integer.parseInt(string.substring(5, 9)));
+  }
+
+  /**
+   * Restore the formatted string into a pair of vertex/edge list and the execution property index.
+   *
+   * @param string the formatted string, for patterns.
+   * @param dag the IR DAG to observe.
+   * @return a pair of vertex/edge list and the execution property key index.
+   */
+  public static Pair<List<Object>, Integer> patternStringToObjAndEPKeyIndex(final String string, final IRDAG dag) {
+    // Formatted into 10 digits: 0:vertex/edge 1-3: patternID 4-5: vertex/edge 6-9:EP Index.
+    final Integer idx = Integer.parseInt(string.substring(0, 1));
+    if (string.length() != 10 || idx != 1) {
+      throw new InvalidParameterException("The metric data should follow the format of "
+        + "[0]: index indicating vertex/edge, [1-3]: id of the pattern, [4-5]: vertex/edge, "
+        + "and [6-9]: EP Key index. Current: " + string);
+    }
+    final IRDAG subDAG = MetricUtils.getSubDAGFromIndex(Integer.parseInt(string.substring(1, 4)));
+    final IRVertex targetVertex = subDAG.getTopologicalSort().get(subDAG.getVertices().size() - 1);
+    if (subDAG.getEdges().stream().anyMatch(e -> e.getSrc().equals(targetVertex))) {
+      throw new MetricException(targetVertex.getId() + " is not the target vertex");
+    }
+
+    final Stream<IRVertex> targetVertexFromDAG = dag.getVertices().stream()
+      .filter(v -> v.toString().equals(targetVertex.toString()))
```

Comment:
I'm not sure that checking the equality using `toString()` makes sense. Two vertices could have the equal `toString()` but could be different instances (have different instance variables). In fact, `toSrtring()` does not guarantee equality.   Do you suppose just checking the same class? then, why don't you use `instanceof` ?

Reply:
Good point I'll fix it

Reply:
Actually, thinking about it, `instanceof` does not check if both classes have the same transform for OperatorVertices or the same source for SourceVertices. That's why I've gone for the toString methods. I could make the code a bit more complex and compare the transform/source classes if you think that it makes it simpler

Reply:
Are you checking `transform` of OperatorVertex? then why don't you use `vertex.getTransform() instanceof`? It is more clear than just using `toString()`. 
And you can also check the source in a similar way: `.getSource() instanceof`?